core/src/Kokkos_View.hpp

```diff
@@ -775,22 +783,102 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
-                  "overload taking a layout object instead.");
+#ifdef KOKKOS_COMPILER_MSVC
+  // MSVC had pack expansion issues with the condition inside the enable_if
+ private:
+  template <class... Args>
+  static constexpr bool msvc_workaround_ctor_condition_1() {
+    size_t num_args = sizeof...(Args);
+    bool are_constructible =
+        (std::is_constructible_v<size_t, Args> && ... && true);
+    return (num_args != rank() + 1) && are_constructible;
+  }
+
+ public:
+#endif
+
+  template <class... Args>
+  View(std::enable_if_t<
+#ifndef KOKKOS_COMPILER_MSVC
+           ((sizeof...(Args)) != rank() + 1) &&
+               (std::is_constructible_v<size_t, Args> && ... && true)
+#else
+           msvc_workaround_ctor_condition_1<Args...>()
+#endif
+               ,
+           const std::string&>
+           arg_label,
+       const Args... args)
+      : View(Impl::ViewCtorProp<std::string>(arg_label), args...) {
+  }
+
+ private:
+  // Special thing for Sacado taking rank()+1 integers, where the last integer
+  // is the FAD dimension
+  template <class... Args, size_t... Idx>
+  static auto view_alloc_from_label_and_integrals(std::true_type,
+                                                  const std::string& arg_label,
+                                                  std::index_sequence<Idx...>,
+                                                  Args... args) {
+    return view_alloc(arg_label, Impl::AccessorArg_t{static_cast<size_t>(
+                                     ((Idx == rank() ? args : 0) + ... + 0))});
+  }
+
+  template <class... Args, size_t... Idx>
+  static auto view_alloc_from_label_and_integrals(std::false_type,
+                                                  const std::string& arg_label,
+                                                  std::index_sequence<Idx...>,
+                                                  Args...) {
+    return view_alloc(arg_label);
+  }
+
+#ifdef KOKKOS_COMPILER_MSVC
+  // Same as above but checking for num_args equal to rank()+1
+  template <class... Args>
+  static constexpr bool msvc_workaround_ctor_condition_2() {
+    size_t num_args = sizeof...(Args);
+    bool are_constructible =
+        (std::is_constructible_v<size_t, Args> && ... && true);
+    return (num_args == rank() + 1) && are_constructible;
+  }
+#endif
+
+ public:
+  template <class... Args>
+  View(std::enable_if_t<
+#ifndef KOKKOS_COMPILER_MSVC
+           ((sizeof...(Args)) == rank() + 1) &&
+               (std::is_constructible_v<size_t, Args> && ... && true)
+#else
+           msvc_workaround_ctor_condition_2<Args...>()
+#endif
+               ,
+           const std::string&>
+           arg_label,
+       const Args... args)
+      : View(view_alloc_from_label_and_integrals(
+                 std::bool_constant<traits::impl_is_customized>(), arg_label,
+                 std::make_index_sequence<sizeof...(Args)>(), args...),
+             args...) {
+  }
```

Comment:
Shouldn't this test for `impl_is_customized` as well?

Reply:
No I don't think so. The problem is that we need the non-ambiguous overload sets. So we would need to write the combination of "is_customized" and "num_args == rank()+1" being true and false. So four ctors in total. This way we have only 2: one which handles the case for num_args != rank()+1 which works for both customized and none-customized. And then we deal with the customized/non-customized inside the ctor for the num_args==rank()+1 case.